crengine/src/chmfmt.cpp

```diff
@@ -11,6 +11,10 @@
 #include "../include/crlog.h"
 #include <chm_lib.h>
 
+#include "../include/fb2def.h"
+#define XS_IMPLEMENT_SCHEME 1
+#include "../include/fb2def.h"
```

Comment:
This looks odd, as we have this comment:
https://github.com/buggins/coolreader/blob/bec2becc7ea8e34daf96a32e19f3da9810751eb9/crengine/include/fb2def.h#L4-L10

But I guess it's the easier way - and also, it will be local to chmfmt.cpp/chmfmt.o and shouldn't cause any issue ?

(The alternative would be to pick the _elementNameTable & co from the _`doc` passed from lvdocview to CHMTOCReader (which has them) and add them to the temporary `doc` LVParseCHMHTMLStream creates - but this looks more complicated - and more work for a format we don't really care about :)

Reply:
Yes, we include in chmfmt.cpp next items:
```cpp
static elem_def_t fb2_elem_table [] = { ... };
static attr_def_t fb2_attr_table [] = { ... };
static ns_def_t fb2_ns_table [] = { ... };
```
Ok, for alternative we must add getters, setters for `lxmlDocBase::_elementNameTable`, `lxmlDocBase::_attrNameTable`, `lxmlDocBase::_nsNameTable` and pass it to temporary `doc` in the `LVParseCHMHTMLStream()` function. Well, if you think that it will be more correct - do it. Later, as usual, I sync the code.

Reply:
I as much as you don't really want to do it, so go ahead with this :)
(It will add 20KB to libcrengine.so, which is not much.)

(Had a quick try with the following, as there is:
```c
/// Copy constructor - copies ID tables contents
lxmlDocBase::lxmlDocBase( lxmlDocBase & doc )
/// creates empty document which is ready to be copy target of doc partial contents
ldomDocument::ldomDocument( ldomDocument & doc )
: lxmlDocBase(doc)
```
but then, it would to do too much and crash when wanting to init styles on this new document... So, let's forget that option :)
```diff
@@ -861,3 +857,3 @@ public:

-ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingName )
+ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingName, ldomDocument * parent_doc )
 {
@@ -903,8 +899,6 @@ ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingNa
     ldomDocument * doc;
-    doc = new ldomDocument();
+    doc = new ldomDocument(*parent_doc);
     doc->setDocFlags( 0 );
-    doc->setNodeTypes(fb2_elem_table);
-    doc->setAttributeTypes(fb2_attr_table);
-    doc->setNameSpaceTypes(fb2_ns_table);

+    printf("new doc blockquote id=%d\n", doc->getElementNameIndex(U"blockquote"));
     ldomDocumentWriterFilter writerFilter(doc, false, HTML_AUTOCLOSE_TABLE);
@@ -1106,3 +1100,3 @@ public:
             }
-            ldomDocument * doc = LVParseCHMHTMLStream( tocStream, defEncodingName );
+            ldomDocument * doc = LVParseCHMHTMLStream( tocStream, defEncodingName, _doc );
             if ( !doc ) {
```

Reply:
Just to get back these 20KB, I might go with:
```diff
--- a/crengine/include/lvtinydom.h
+++ b/crengine/include/lvtinydom.h
@@ -1300,2 +1300,4 @@ public:
     void setNameSpaceTypes( const ns_def_t * ns_scheme );
+    // set node/attribute/namespace types by copying them from an other document
+    void setAllTypesFrom( lxmlDocBase * d );

diff --git a/crengine/src/chmfmt.cpp b/crengine/src/chmfmt.cpp
index 1f4f23f6..4b7a80ad 100644
--- a/crengine/src/chmfmt.cpp
+++ b/crengine/src/chmfmt.cpp
@@ -7,6 +7,2 @@

-#include "../include/fb2def.h"
-#define XS_IMPLEMENT_SCHEME 1
-#include "../include/fb2def.h"
-
 #define DUMP_CHM_DOC 0
@@ -861,3 +857,3 @@ public:

-ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingName )
+ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingName, ldomDocument * parent_doc )
 {
@@ -905,5 +901,3 @@ ldomDocument * LVParseCHMHTMLStream( LVStreamRef stream, lString32 defEncodingNa
     doc->setDocFlags( 0 );
-    doc->setNodeTypes(fb2_elem_table);
-    doc->setAttributeTypes(fb2_attr_table);
-    doc->setNameSpaceTypes(fb2_ns_table);
+    doc->setAllTypesFrom(parent_doc);

@@ -1106,3 +1100,3 @@ public:
             }
-            ldomDocument * doc = LVParseCHMHTMLStream( tocStream, defEncodingName );
+            ldomDocument * doc = LVParseCHMHTMLStream( tocStream, defEncodingName, _doc );
             if ( !doc ) {
diff --git a/crengine/src/lvtinydom.cpp b/crengine/src/lvtinydom.cpp
index e5e69b78..9e0fa14e 100644
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -5119,2 +5119,23 @@ void lxmlDocBase::setNameSpaceTypes( const ns_def_t * ns_scheme )

+// set node/attribute/namespace types by copying them from an other document
+void lxmlDocBase::setAllTypesFrom( lxmlDocBase * d )
+{
+    // Simpler to just serialize and deserialize them all
+    SerialBuf buf(0, true);
+    d->_elementNameTable.serialize( buf );
+    buf << d->_nextUnknownElementId;
+    d->_attrNameTable.serialize( buf );
+    buf << d->_nextUnknownAttrId;
+    d->_nsNameTable.serialize( buf );
+    buf << d->_nextUnknownNsId;
+
+    buf.setPos(0);
+    _elementNameTable.deserialize( buf );
+    buf >> _nextUnknownElementId;
+    _attrNameTable.deserialize( buf );
+    buf >> _nextUnknownAttrId;
+    _nsNameTable.deserialize( buf );
+    buf >> _nextUnknownNsId;
+}
+
 void lxmlDocBase::dumpUnknownEntities( const char * fname )
```
if that's ok with you.
I checked that `printf("chm blockquote id=%d\n", doc->getElementNameIndex(U"blockquote"));` gives 26 (as with your code) instead of 512, so I guess it should fix your original auto close tags issue too.

Reply:
For info because it's a bit related, I've added this to some other cleanup work of mine (soon to be PR'ed):
```diff
--- a/crengine/src/lvdocview.cpp
+++ b/crengine/src/lvdocview.cpp
@@ -4651,6 +4651,17 @@ void LVDocView::createEmptyDocument() {
     m_doc->setNodeTypes(fb2_elem_table);
     m_doc->setAttributeTypes(fb2_attr_table);
     m_doc->setNameSpaceTypes(fb2_ns_table);
+    // Note:
+    // All book formats parsing start with this, and get these fb2def.h
+    // elements defined and included in the document ids maps. This allows
+    // all our specific XHTML DOM element handling in ldomDocumentWriter
+    // using "==el_body", "==el_head" to match on books.
+    // But secondary XML files (.opf, .ncx...) get parsed with
+    // LVParseXMLStream(), which will not inject these known ids maps
+    // and will start a DOM with empty maps: met elements and attributes
+    // will get IDs starting from 512 (even if named <body> or <head>),
+    // and won't trigger these specific elements handling where we
+    // use these < 512 IDs (el_body, el_head...), which is good!
 }
```

Reply:
It seems to work. Who makes the commit?  Me and then you copy or you and then I copy?

Reply:
As you wish.
I'll be making a PR later today with it among others, I'll just tag this one "(Upstream)" if you make a PR before me.
But if you wait for my PR to pick it - you'll have 20 other commits to pick to keep in sync, so it depends on when you want to do that painful koreader-sync work :/ There may also be some risky commits, so you might want to wait a bit before syncing.